Issue/13326 scan full screen loading & error handling

ashiagr opened the pull request:
Parent #13326

This PR displays full screen loading and errors on fetch scan and start scan requests. 

Full screen loading/ errors
Loading (0) | No Network (1) | Generic Request Failed (3)  | Start Scan Request Failed (4) 
---------|-------------|------------------------|--------------------------
![loading_scan](https://user-images.githubusercontent.com/1405144/106376887-24aaaa80-63bf-11eb-91d3-5780701d90dd.png)| ![no-network-full screen](https://user-images.githubusercontent.com/1405144/106376942-78b58f00-63bf-11eb-89da-d5ed9edb477e.jpg) |![generic_request_failed](https://user-images.githubusercontent.com/1405144/106376911-45730000-63bf-11eb-87df-e4b5e3cf8fd8.png)|![Something went wrong_start_scan](https://user-images.githubusercontent.com/1405144/106376931-5c195700-63bf-11eb-9aad-f96361af859a.png)

Snack bars
No Network (5)| Generic Request Failed (6)
-----------| -----------------------
![no network snackbar](https://user-images.githubusercontent.com/1405144/106376981-b74b4980-63bf-11eb-978d-83c067736cde.jpg)|![generic_request_failed_snackbar](https://user-images.githubusercontent.com/1405144/106376963-9be03e80-63bf-11eb-80f5-76687b31af0b.png)

### To test

#### Loading (0), No Network (1)

1. Fresh install the app and logon to wp.com account with site having scan capability
2. Switch to site with scan capability
3. Enable airplane mode
4. Click Scan menu
5. Notice that fullscreen "Loading scan" is shown, followed by full screen "No network available" message
6. Disable airplane mode
7. Click retry
8. Notice that scan content is shown
 
------------------------------------------------------------

#### No network snackbar (5)

1. Continuing from above, while being on the scan screen, enable airplane mode
2. Click scan button
3. Notice that a snackbar is shown with  no network available message

------------------------------------------------------------

#### Generic Request Failed (3), Start Scan Request Failed (4), Generic Request Failed snackbar (6)
These states can be tested by using
`ScanViewModel` tests:

Generic Request Failed (3)
`given fetch scan fails, when scan state fetched over empty scan state, then request failed ui state shown`
`given request failed error ui state, when contact support is clicked, then contact support screen is shown`

Start Scan Request Failed (4)
`given scan start request fails, when scan button is clicked, then request failed state is shown`
`given start scan request failed error state, when contact support is clicked, then contact support shown`

Generic Request Failed snackbar (6)
`given fetch scan state fails, when scan state fetched over last scan state, then request failed msg shown`

Notes: 

1. Start scan request fail displays a full screen error with contact support button, message string is copied from the iOS app. 
2. I plan to implement pull-to-refresh which will enable users to retry fetch scan state on full screen errors. 

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Refactor: Update recycler view properties in initRecyclerView

ashiagr committed:
Refactor: Update Content to ContentUiState

ashiagr committed:
Add ActionableEmptyView

ashiagr committed:
Add full screen loading ui state

ashiagr committed:
Add fetch scan state error handling

ashiagr committed:
Tests: Add tests for fetch scan state error handling

ashiagr committed:
Add start scan state error handling

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/ScanViewModel.kt:
```diff
@@ -262,7 +291,41 @@ class ScanViewModel @Inject constructor(
-        data class Content(val items: List<JetpackListItemState>) : UiState()
+    sealed class UiState {
+        object FullScreenLoadingUiState : UiState() {
+            @DrawableRes val image = R.drawable.img_illustration_empty_results_216dp
```
🔍 Nitpick: The illustration looks a bit weird on "loading" screen. I'd consider showing a progress instead of the "result empty" illustration. Wdyt?


P.S. Or a skeleton with a Shimmer effect :).

ashiagr replied:
~~I'll add a skeleton with a Shimmer effect.~~

ashiagr replied:
As discussed, ended up using progress bar in d9ea5d2

malinajirka commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -1061,6 +1061,14 @@
     <string name="scan_history">Scan History</string>
     <string name="scan_history_menu_title">History</string>
 
+    <string name="scan_loading_title">Loading scan</string>
```
🔍 Nitpick: We might want to consider adding "..." at the end of this string resource. Wdyt?

ashiagr replied:
Nitpick accepted, added '...' in cf7dfd6.

malinajirka commented on WordPress/src/test/java/org/wordpress/android/ui/jetpack/scan/ScanViewModelTest.kt:
```diff
@@ -89,6 +93,18 @@ class ScanViewModelTest : BaseUnitTest() {
+        whenever(scanStore.getScanStateForSite(site)).thenReturn(null)
+        val uiStates = init().uiStates
+
+        val state = uiStates.first() as FullScreenLoadingUiState
```
🔍 Nitpick: It seems that this tests doesn't have just a single reason to fail. We might want to consider splitting it into two tests.
1. One that check that `uiStates.first()` is instance of `FullScreenLoadingUiState`
2. One that checks that content of `FullScreenLoadingUiState` is what we expect

Same applies to some of the other tests, wdyt?

ashiagr replied:
It confuses me always, split done in 273866b.

malinajirka commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -1061,6 +1061,14 @@
+    <string name="scan_loading_title">Loading scan</string>
+    <string name="scan_no_network_title" translatable="false">@string/no_network_title</string>
+    <string name="scan_no_network_subtitle" translatable="false">@string/no_network_message</string>
+    <string name="scan_request_failed_title" translatable="false">Something went wrong</string>
```
⚠️ It seems this string uses "translatable=false" attribute, even though it contains a translation and not just a reference. Wdyt?

ashiagr replied:
Good catch. After I added it, I searched for "Something went wrong" in the `strings.xml `as I remembered I saw it earlier. But it wasn't the same and had an extra string added to it, then I forgot to remove "translatable=false" 🤦‍♀️.

In cf7dfd6.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/ScanFragment.kt:
```diff
@@ -66,8 +71,14 @@ class ScanFragment : Fragment(R.layout.scan_fragment) {
             { uiState ->
-                if (uiState is Content) {
-                    refreshContentScreen(uiState)
+                when (uiState) {
```
❓ Do we need to update visibility of these views? Or we don't need to do that, because the RV is always empty when a loading/error fullscreen view is shown?

ashiagr replied:
We don't need to do that, because the RV is always empty when a loading/error fullscreen view is shown.

malinajirka reviewed (changes requested):
Thanks @ashiagr! Great job ;). It looks good overall, I've left one blocking comment ("translatable=false") and a few nitpicks.

🔍 Nitpick: When I click on "retry" when my internet connection is still disabled, there is no loading indicator  ->  it seems like the button doesn't work. We usually add "500ms" delay in such scenarios so the loading is displayed for a bit.

ashiagr committed:
Remove translatable = false, add "..." in loading scan string

ashiagr committed:
Split tests to follow single reason to fail principle

ashiagr committed:
Replace empty view having loading label with progress bar, add retry delay

ashiagr commented:
Thanks for the review @malinajirka! I've addressed your comments and answered questions.

>  Nitpick: When I click on "retry" when my internet connection is still disabled, there is no loading indicator -> it seems like the button doesn't work. We usually add "500ms" delay in such scenarios so the loading is displayed for a bit.

Added retry delay in d9ea5d2.

Ready for another round.

ashiagr committed:
Remove unused string

malinajirka reviewed (approved):
LGTM, thanks! ;)
